extra_data/file_access.py

```diff
@@ -200,6 +212,29 @@ def _read_data_sources(self):
 
         return frozenset(control_sources), frozenset(instrument_sources)
 
+    def _guess_valid_trains(self):
+        # File format version 1.0 includes a flag which is 0 if a train ID
+        # didn't come from the time server. We use this to skip bad trains,
+        # especially for AGIPD.
+        # Older files don't have this flag, so this tries to estimate validity.
+        # The goal is to have a monotonic sequence within the file with the
+        # fewest trains skipped.
+        train_ids = self.train_ids
+        flag = np.ones_like(train_ids, dtype=bool)
+
+        for ix in np.nonzero(train_ids[1:] <= train_ids[:-1])[0]:
+            # train_ids[ix] > train_ids[ix + 1]
+            invalid_before = train_ids[:ix+1] > train_ids[ix+1]
+            invalid_after = train_ids[ix+1:] < train_ids[ix]
+            # Which side of the downward jump in train IDs would need fewer
+            # train IDs invalidated?
+            if np.count_nonzero(invalid_before) < np.count_nonzero(invalid_after):
```

Comment:
This considers each 'step down' separately to decide whether to invalidate trains before the step or after. In the cases we've had most problems with, there's just one train ID which is obviously out of sequence, and this should easily eliminate it.

But this could go wrong in other scenarios - e.g. if you had train IDs interleaved from two sequences, one higher and one lower, considering each step from the higher to the lower separately will result in a mess where you preserve the first half of the lower sequence and the second half of the higher one, when the 'right' answer is probably to preserve all of one sequence. I haven't thought of a way to consider all of step-downs together and efficiently find train IDs to mark as invalid.

This also assumes that train IDs should always be recorded in increasing order, and it's reasonable to skip some data if they're not. It's conceivable that you might want the 'bad' data anyway, e.g. if you don't care about matching it up with other sources. But we've always flagged this as invalid with `extra-data-validate`, and the issues we've seen with AGIPD would generally result in dropping one train in a run of many thousands.